### PR TITLE
[X] TypeConverter on Grid Col/Row Definitions

### LINKED ISSUE #4997: [Enhancement] Less verbose grid layout definition
## Summary
Current grid layout XAML definition too verbose and hard to maintain, the grid structure definition can be simplified with an alternative way to specify it with the properties.

## API Changes

Add new properties to grid: 
 *  Cols  - string definition of the columns
 *  Rows - string definitions of the rows
 *  Pos - attached property to simplify Column/Row in a format x,[y,[w,[h]]]
Examples:

Cols="a,*,3:a"  
```
		<Grid.ColumnDefinitions>
			<ColumnDefinition Width="Auto" />
			<ColumnDefinition Width="*" />
			<ColumnDefinition Width="Auto" />
			<ColumnDefinition Width="Auto" />
			<ColumnDefinition Width="Auto" />
		</Grid.ColumnDefinitions>
```

Rows="10,a,3:*"
```
		<Grid.RowDefinitions>
			<RowDefinition Height="10" />
			<RowDefinition Height="Auto" />
			<RowDefinition Height="*" />
			<RowDefinition Height="*" />
			<RowDefinition Height="*" />
		</Grid.RowDefinitions>
```

Pos="1,2,4,3"
```
		<... Grid.Column="1"   
                     Grid.Row="2" 
                     Grid.ColumnSpan="3" 
                     Grid.RowSpan="4" 
                    ...>
```
## API Changes
Possible implementation:
```
public class Grid 
	{
...
		#region Position
		public static readonly BindableProperty PosProperty =
			BindableProperty.CreateAttached("Pos",
				typeof(GridLocation),
				typeof(Grid),
				null,
				propertyChanged: OnPositionChanged);

		private static void OnPositionChanged(BindableObject bindable, object oldValue, object newValue)
		{
			if (newValue is GridLocation pos)
			{
				bindable.SetValue(RowProperty, pos.Row);
				bindable.SetValue(ColumnProperty, pos.Column);
				bindable.SetValue(RowSpanProperty, pos.RowSpan);
				bindable.SetValue(ColumnSpanProperty, pos.ColumnSpan);
			}
		}

		public static GridLocation GetPos(BindableObject bindable)
		{
			return (GridLocation)bindable.GetValue(PosProperty);
		}

		public static void SetPos(BindableObject bindable, GridLocation value)
		{
			bindable.SetValue(PosProperty, value);

			OnPositionChanged(bindable, null, value);
		}
		#endregion

		#region Columns
		public static readonly BindableProperty ColsProperty =
		 BindableProperty.Create(nameof(Cols),
			 typeof(GridDefinition),
			 typeof(Grid),
			 null,
			 propertyChanged: OnUpdateGridColumns);

		public GridDefinition Cols
		{
			get => (GridDefinition)GetValue(ColsProperty);
			set => SetValue(ColsProperty, value);
		}

		private static void OnUpdateGridColumns(BindableObject bindable, object oldValue, object newValue)
		{
			if (newValue is GridDefinition def)
			{
				if (bindable is Grid grid)
				{
					var collection = new ColumnDefinitionCollection();
					foreach (var c in def.Grids)
					{
						collection.Add(new ColumnDefinition { Width = c });
					}
					grid.ColumnDefinitions = collection;
				}
			}
		}
		#endregion

		#region Rows
		public static readonly BindableProperty RowsProperty =
		  BindableProperty.Create(nameof(Rows),
			  typeof(GridDefinition),
			  typeof(Grid),
			  null,
			  propertyChanged: OnUpdateGridRows);
		public GridDefinition Rows
		{
			get => (GridDefinition)GetValue(RowsProperty);
			set => SetValue(RowsProperty, value);
		}
		private static void OnUpdateGridRows(BindableObject bindable, object oldValue, object newValue)
		{
			if (newValue is GridDefinition def)
			{
				if (bindable is Grid grid)
				{
					var collection = new RowDefinitionCollection();
					foreach (var c in def.Grids)
					{
						collection.Add(new RowDefinition { Height = c });
					}
					grid.RowDefinitions = collection;
				}
			}
		}
		#endregion 
	}


	[Xamarin.Forms.Xaml.TypeConversion(typeof(GridDefinition))]
	public class GridDefinitionTypeConverter : TypeConverter
	{
		public override object ConvertFromInvariantString(string param)
		{

			return GridDefinition.ValueOf(param);
		}
	}

	[TypeConverter(typeof(GridDefinitionTypeConverter))]
	public class GridDefinition
	{
		private IList<GridLength> _grids = new List<GridLength>();
		public IList<GridLength> Grids
		{
			get => new List<GridLength>(_grids);
			private set => _grids = value;
		}


		public GridDefinition()
		{
		}

		private readonly static IDictionary<string, GridDefinition> _cached = new Dictionary<string, GridDefinition>();
		internal static object ValueOf(string param)
		{
			if (_cached.TryGetValue(param, out var result))
			{
				return result;
			}
			result = new GridDefinition();

			var grids = new List<GridLength>();
			foreach (var i in param.Split(','))
			{
				string def = i?.Trim()?.ToLower();
				GridLength length = GridLength.Auto;
				int multi = 1;

				if (def.Contains(":"))
				{
					if (!int.TryParse(def.Split(':')[1], out multi))
					{
						throw new ArgumentException($"Invalid grid definition value {param}");
					}
					def = def.Split(':')[0];
				}

				if (def.Contains("*"))
				{
					double value = 1.0;
					if (!def.StartsWith("*"))
					{
						var digits = def.Trim('*');
						if (!double.TryParse(digits, out value))
						{
							throw new ArgumentException($"Invalid grid definition value {param}");
						}
					}
					length = new GridLength(value, GridUnitType.Star);
				}
				else
				{

					if (def != "a" && def != "auto")
					{
						if (!double.TryParse(def, out double value))
						{
							throw new ArgumentException($"Invalid grid definition value {param}");
						}
						length = new GridLength(value, GridUnitType.Absolute);
					}
				}

				while (multi > 0)
				{
					grids.Add(length);
					multi--;
				}

			}
			result.Grids = grids;
			_cached[param] = result;
			if (_cached.Count > 500)
			{
				_cached.Clear();
			}

			return result;
		}
	}

	[Xamarin.Forms.Xaml.TypeConversion(typeof(GridLocation))]
	public class GridLocationTypeConverter : TypeConverter
	{
		public override object ConvertFromInvariantString(string param)
		{
			return GridLocation.ValueOf(param);

		}
	}

	[TypeConverter(typeof(GridLocationTypeConverter))]
	public class GridLocation
	{
		public int Row { get; private set; } = 0;
		public int Column { get; private set; } = 0;
		public int RowSpan { get; private set; } = 1;
		public int ColumnSpan { get; private set; } = 1;

		public GridLocation(int x) : this(x, 0) { }

		public GridLocation(int x, int y) : this(x, y, 1) { }


		public GridLocation(int x, int y, int w) : this(x, y, w, 1) { }


		public GridLocation(int x, int y, int w, int h)
		{
			Row = y;
			Column = x;
			RowSpan = h;
			ColumnSpan = w;
		}

		private readonly static IDictionary<string, GridLocation> _cached = new Dictionary<string, GridLocation>();
		internal static object ValueOf(string param)
		{
			if (_cached.TryGetValue(param, out var result))
			{
				return result;
			}
			IList<int> values = new List<int>();
			foreach (var d in param.Split(','))
			{
				if (!int.TryParse(d.Trim(), out var val))
				{
					throw new ArgumentException($"Invalid location definition value {param}");
				}
				values.Add(val);
			}

			if (values.Count > 4 || values.Count == 0)
			{
				throw new ArgumentException($"Invalid location definition value {param}");
			}
			GridLocation location = null;
			switch (values.Count)
			{
				case 1: location = new GridLocation(values[0]); break;
				case 2: location = new GridLocation(values[0], values[1]); break;
				case 3: location = new GridLocation(values[0], values[1], values[2]); break;
				case 4: location = new GridLocation(values[0], values[1], values[2], values[3]); break;
			}
			_cached[param] = location;
			if (_cached.Count > 500)
			{
				_cached.Clear();
			}
			return location;
		}
	}
```

### LINKED ISSUE #7302: Proposal: Simplify Grid Column and Row Definitions
## Summary
WinUI is adding a simplified syntax to setting column and row definitions on Grid. Xamarin.Forms should add this too.

### Current syntax
```xml
<Grid>
    <Grid.ColumnDefinitions>
          <ColumnDefinition Width="1*" />
          <ColumnDefinition Width="2*" />
          <ColumnDefinition Width="Auto" />
          <ColumnDefinition Width="*" />
          <ColumnDefinition Width="300" />
    </Grid.ColumnDefinitions>
    <Grid.RowDefinitions>
          <RowDefinition Height="1*" />
          <RowDefinition Height="Auto" />
          <RowDefinition Height="25" />
          <RowDefinition Height= "14" />
          <RowDefinition Height="20" />
    </Grid.RowDefinitions>
</Grid>
```
### Proposed Syntax
The same functionality as above with the proposed succinct syntax is shown below:
```xml
<Grid ColumnDefinitions="1*, 2*, Auto, *, 300" RowDefinitions="1*, Auto, 25, 14, 20"/>
```
Full spec described here: https://github.com/microsoft/microsoft-ui-xaml/issues/673 (design is currently being finalized)

## API Changes
Follow the same specification implemented by WinUI (probably no new APIs, unless it needs a new type-converter)

## Intended Use Case
Developer productivity

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###
Simplify Grid Column and Row Definitions by using markup extensions.

### Issues Resolved ### 

- fixes #7302
- closes #4997

### API Changes ###

Added:
 - ColumnDefinitionCollectionTypeConverter
 - RowDefinitionCollectionTypeConverter

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### PR Checklist ###

- [x] Targets the correct branch
- [x] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by jamesmontemagno** [comment]
YES YES YES YES YES YES YES

File: Xamarin.Forms.Core/ColumnDefinitionCollectionTypeConverter.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Xamarin.Forms
+{
+	[Xaml.TypeConversion(typeof(ColumnDefinitionCollection))]
+	public class ColumnDefinitionCollectionTypeConverter : TypeConverter
+	{
+		public override object ConvertFromInvariantString(string value)
+		{
+			if (value != null) {
+				var lengths = value.Split(',');
```
Comment (dotMorten) [inline Xamarin.Forms.Core/ColumnDefinitionCollectionTypeConverter.cs:11]: To follow common xaml, both spaces and commas can be used. See the original spec that the issue linked to .
Also see the Thickness Converter.

**Comment by StephaneDelcroix** [comment]
@dotMorten at this point, we have no precedent on using spaces as list separator in XAML values. 

What you see on Thickness, and some other converters, is us reusing the same converter for css value parsing. CSS list separator are actually spaces, but the order of arguments differs (rtlb vs trbl).

We could start discussing supporting different list separators in a different thread, but at this time I see very little traction or demand for this.

**Comment by dotMorten** [comment]
Makes sense. Just another of those differences between dialects that shouldn't be I guess

**Comment by MitchBomcanhao** [comment]
@dotMorten how would the following column definition be converted to this new format?
is this supported?
```
<Grid.ColumnDefinitions>
   <ColumnDefinition Width="Auto" />
   <ColumnDefinition Width="{x:Static local:AppConstants.PadHalf}" />
</Grid.ColumnDefinitions>
```

**Comment by dotMorten** [comment]
@MitchBomcanhao It wouldn't. Advanced cases like this, you'd still need to use the full syntax.

**Comment by MitchBomcanhao** [comment]
@dotMorten thanks for confirming that.

**Comment by EmilAlipiev** [comment]
Does this not cause slow rendering if we use converter to retrieve column and row definitions?

**Comment by StephaneDelcroix** [comment]
> Does this not cause slow rendering if we use converter to retrieve column and row definitions

TypeConverter is mainly used at the time the XAML is parsed, and parsing the MarkupExtension is no slower than parsing the equivalent Xaml nodes

**Pull request merged**